### PR TITLE
Refactoring change flag

### PR DESCRIPTION
**Pull request by hennevogel**

### CODE REVIEW

**Comment by hennevogel** [comment]
This should get a review

**Comment by adrianschroeter** [comment]
you run into situations where you have the same architectures multiple times (because they are used in multiple repos) and in some times you need them just uniq.

There is just no direct dependency of architectures to project. If it is needed for the webui views, it belongs to some webui helper IMHO.

**Comment by hennevogel** [comment]
I still don't get it what your problem is sorry.

**Comment by hennevogel** [comment]
Okay we talked about this in the office. Adrian thinks that we should use this relationship in  `Project` 

```
has_many :architectures, -> { order("position") }, :through => :repository_architectures
```

Because there is no direct relationship between `Project` and `Architecture`. And I'm now using this in the `has_flags` mixin. 

Please review @openSUSE/open-build-service please

**Comment by mdeniz** [comment]
What's wrong with that relationship? I don't get it... Is just a usual activerecord relationship "through" other model.
@adrianschroeter can you explain?

**Comment by adrianschroeter** [comment]
For one I find code like this:

update_project.architecture.each ...

misleading, since it makes the impression that a project has a direct relationship to architectures. We do not have this anywhere else IMHO.

Atm. you get the same architectures multiple times via that (because you them used in multiple repos). K, you can already make them uniq, but the design will break when we extend the architectures later on with
optional attributes. You won't see these at all here.

I like to propose to this either via:

 1) move this code to some view/helper, since it makes only sense for these specific flag views, so it does not belong to the generic model.

or

 2) don't call it architectures, which creates the wrong assumption that a project has architectures, but deliver the data via some method called like "Project.uniq_architecture_list_of_all_repositories" ... or something which is a bit shorter maybe ;)

**Comment by adrianschroeter** [comment]
k, then just merge it for now if you both disagree, the working UI for flags is more important then the style question.

**Comment by bgeuken** [comment]
I found the naming strange as well. Something like enabled, available or *** architectures seems to make more sense to me. Though that seems to be out of scope of this PR.

And since I overheard your discussion this afternoon, wondering why the relation got merged few weeks ago... Maybe we should adjust our review process to ensure that we have a consens about certain kind of changes. Introducing rubocop rules that enforce certain coding styles would be one thing that comes to my mind here.

**Comment by bgeuken** [comment]
And regarding the PR... Why do we need a GetFlags module here?:(

File: src/api/app/mixins/get_flags.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    the_flags = {}
+
+    # FIXME: A helper called in a model...
+    return the_flags unless FlagHelper::TYPES.has_key?(flag_type.to_s)
+
+    # [nil] is for no architecture/repository
+    ['all'].concat(self.repositories.map{|repo| repo.name}).each do |repository|
+      the_flags[repository] = []
+      ['all'].concat(self.architectures.uniq).each do |architecture|
+        arch = ( architecture == 'all' ) ? nil : architecture.name
```
Comment (bgeuken) [inline src/api/app/mixins/get_flags.rb]: Is it possible that self.architectures includes an arch called "all"?

File: src/api/app/mixins/get_flags.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        architecture = nil if architecture == 'all'
```
Comment (bgeuken) [inline src/api/app/mixins/get_flags.rb]: This feels like a duplication of the line above. Could you merge both lines?

File: src/api/app/mixins/get_flags.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  #        this happens here is because rendering the XML does ignore the defaults.
+  # Returns a hash of arrays, sorted by repository.
+  # The arrays contain Flag objects of type, sorted by architecture.
+  # Like:
+  # {all: [Flag, Flag-i586, Flag-x86_64], 13.2: [Flag, Flag-i585, Flag-x86_64]}
+  def get_flags(flag_type)
+    the_flags = {}
+
+    # FIXME: A helper called in a model...
+    return the_flags unless FlagHelper::TYPES.has_key?(flag_type.to_s)
```
Comment (bgeuken) [inline src/api/app/mixins/get_flags.rb]: Do we expect this to happen? It feels like this would be a bug, wouldn't it?

File: src/api/app/mixins/get_flags.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    # [nil] is for no architecture/repository
+    ['all'].concat(self.repositories.map{|repo| repo.name}).each do |repository|
+      the_flags[repository] = []
+      ['all'].concat(self.architectures.uniq).each do |architecture|
+        arch = ( architecture == 'all' ) ? nil : architecture.name
+        architecture = nil if architecture == 'all'
+        repo = ( repository == 'all' ) ? nil : repository
+        flag = self.flags.with_types(flag_type).with_repositories(repo).with_architectures(arch).first
```
Comment (bgeuken) [inline src/api/app/mixins/get_flags.rb]: Wouldn't a where(architecture: architecture) simpler? AFAIK this would even work with arch == nil

File: src/api/app/models/flag.rb [hunk 5]
```diff
@@ -37,9 +37,9 @@ def validate_duplicates
 
   scope :with_types, ->(type) { where(flag: type) }
   scope :with_repositories, ->(repo_name) { where(repo: repo_name) }
```
Comment (bgeuken) [inline src/api/app/models/flag.rb]: I know, this was already there, but this scope seems to obfuscate the code more than it would spare use characters, no?

  ``` .with_repositories(repo) ```
vs
  ``` .where(repo: repo.name) ```

File: src/api/app/mixins/webui/has_flags.rb [hunk 6]
```diff
@@ -40,11 +42,14 @@ def remove_flag
 
     @flag = Flag.find(params[:flag])
     main_object.flags.destroy(@flag)
+    @flag = @flag.dup
```
Comment (bgeuken) [inline src/api/app/mixins/webui/has_flags.rb]: Why is that?

File: src/api/app/models/flag.rb [hunk 7]
```diff
@@ -123,6 +138,12 @@ def fullname
     ret
   end
 
+  def arch
+    architecture_id.blank? ? '' : architecture.name
+  end
 
+  def main_object
+    self.package ? self.package : self.project
```
Comment (bgeuken) [inline src/api/app/models/flag.rb]: self.package || self.project

File: src/api/app/models/flag.rb [hunk 8]
```diff
@@ -123,6 +138,12 @@ def fullname
     ret
   end
 
+  def arch
+    architecture_id.blank? ? '' : architecture.name
```
Comment (bgeuken) [inline src/api/app/models/flag.rb]: architecture.try(:name).to_s

**Comment by hennevogel** [comment]
@bgeuken because we need to share this between project & package controller :-/

File: src/api/app/mixins/get_flags.rb (hunk 1 above)
Comment (hennevogel) [inline src/api/app/mixins/get_flags.rb]: nope it's not

File: src/api/app/mixins/get_flags.rb (hunk 2 above)
Comment (hennevogel) [inline src/api/app/mixins/get_flags.rb]: I don't see how, below I need to somehow distinguish between the name and the object.

File: src/api/app/mixins/get_flags.rb (hunk 4 above)
Comment (hennevogel) [inline src/api/app/mixins/get_flags.rb]: This is a `where` scope in flag. And it does work with `nil` (that's why I need to set it above)

File: src/api/app/mixins/webui/has_flags.rb (hunk 6 above)
Comment (hennevogel) [inline src/api/app/mixins/webui/has_flags.rb]: Because I need to render a new record with the same values

File: src/api/app/models/flag.rb (hunk 5 above)
Comment (hennevogel) [inline src/api/app/models/flag.rb]: Yeah true, I basically introduced this before to be used in views where I don't really like to use `.where`. I'll change it back.

File: src/api/app/mixins/get_flags.rb (hunk 1 above)
Comment (bgeuken) [inline src/api/app/mixins/get_flags.rb]: Then why do you add an "all" arch and replace it with nil one line below? Doesn't seem to have any effect then

File: src/api/app/mixins/get_flags.rb (hunk 3 above)
Comment (hennevogel) [inline src/api/app/mixins/get_flags.rb]: No we don't but we don't validate this ENUM anywhere...

File: src/api/app/mixins/get_flags.rb (hunk 1 above)
Comment (hennevogel) [inline src/api/app/mixins/get_flags.rb]: before it was

```
[nil].concat....
```

and I found this obfuscates what's going on even more...

**Comment by hennevogel** [comment]
And the `architectures` relationship is now distinct. Anything else? :-)

**Comment by hennevogel** [comment]
Oh and @openSUSE/open-build-service when I say **review** I mean the functionality should be as in 2.6, not only code review.

File: src/api/app/mixins/get_flags.rb (hunk 3 above)
Comment (bgeuken) [inline src/api/app/mixins/get_flags.rb]: Wouldn't it then be better to let it fail miserably so we are getting aware of such bugs early?

File: src/api/app/mixins/get_flags.rb [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    return the_flags unless FlagHelper::TYPES.has_key?(flag_type.to_s)
+
+    # [nil] is for no architecture/repository
+    ['all'].concat(self.repositories.map{|repo| repo.name}).each do |repository|
+      the_flags[repository] = []
+      ['all'].concat(self.architectures.uniq).each do |architecture|
+        arch = ( architecture == 'all' ) ? nil : architecture.id
+        architecture = nil if architecture == 'all'
+        repo = ( repository == 'all' ) ? nil : repository
+        flag = self.flags.where(flag: flag_type).where(repo: repo).where(architecture_id: arch).first
```
Comment (bgeuken) [inline src/api/app/mixins/get_flags.rb]: How about this?

```ruby
# Nil is a placeholder for "all" architectures
[nil].concat(self.architectures.uniq).each do |architecture|
flag = self.flags.where(flag: flag_type).where(repo: repo).where(architecture: architecture).first
```

I don't like the [nil] thing either. But it seems easier to read to me than swapping around nil and "all" like before. And "all" isn't actually used.

File: src/api/app/mixins/get_flags.rb (hunk 1 above)
Comment (bgeuken) [inline src/api/app/mixins/get_flags.rb]: At least I find this confusing. It looks like it's "all", but actually it's nil.

**Comment by bgeuken** [comment]
"Oh and @openSUSE/open-build-service when I say review I mean the functionality should be as in 2.6, not only code review"

Pull request => code review
QA => testing of functionality

Not saying that a reviewer shouldn't verify that the proposed code does what it should. But if you are unsure about your changes, you might better test them again;-)

And I sadly don't have a 2.6 instance. Maybe it would have been wise to revert the broken flag commit...

**Comment by hennevogel** [comment]
I've tested these changes but I've also tested my changes before and missed on edge case so I want someone else to look over this now.

File: src/api/app/mixins/get_flags.rb (hunk 9 above)
Comment (hennevogel) [inline src/api/app/mixins/get_flags.rb]: Okay I'll change it back

**Comment by bgeuken** [comment]
* All repo column is not using remote, the others do
* Changes in arch row doesn't change all fields in the arch column (ui only, the meta file is correct)

The created meta file seems to match with the ui now. Though it might be that the ordering of the enabled / disabled elements is wrong. Assuming that the ordering matters for the backend:
```
  <debuginfo>
    <enable arch="i586"/>
    <enable arch="i586" repository="home_adrian_ProtectionTest_repo"/>
    <disable/>
    <disable arch="x86_64" repository="SourceprotectedProject_repo"/>
  </debuginfo>
``

**Comment by hennevogel** [comment]
> All repo column is not using remote, the others do

All `All` colums are not using remote as they change more flags so a reload is necessary. 

> Changes in arch row doesn't change all fields in the arch column (ui only, the meta file is correct)

They do now.

Anyone up for a second review @openSUSE/open-build-service?

**Comment by hennevogel** [comment]
So can we merge this now @openSUSE/open-build-service?

**Comment by bgeuken** [comment]
I briefly looked through it again and it seems safe to merge. Though it would be could if a second person could review this

**Comment by hennevogel** [comment]
@mdeniz, @ChrisBr can you look at this please?

**Comment by bgeuken** [comment]
\*Merging\*
For the sake of getting this deployed:-) and after discussing this today.

Though please update the has_children method in a separate PR. Thanks:-)
